src/librustc_mir/interpret/terminator.rs

```diff
@@ -449,6 +461,25 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         }
     }
 
+    /// Evaluate a const function where all arguments (if any) are zero-sized types.
+    /// The evaluation is memoized thanks to the query system.
+    fn eval_const_fn_call(
+        &mut self,
+        gid: GlobalId<'tcx>,
+        ret: Option<(PlaceTy<'tcx, M::PointerTag>, mir::BasicBlock)>,
+    ) -> InterpResult<'tcx> {
+        trace!("eval_const_fn_call: {:?}", gid);
+
+        let place = self.const_eval_raw(gid)?;
+        let dest = ret.ok_or_else(|| err_ub!(Unreachable))?.0;
```

Comment:
This looks wrong to me. It is not UB to *call* a function that cannot return. It is just UB to *return* from such a function.

Reply:
Isn't this equivalent in the case of zero arg const fns? (Unless we consider nontermination via `loop`, which is fine by me if we abort early on that).

Reply:
I don't think it is fine to claim that a program that runs `loop {}` has UB.

Reply:
true... (well, it's true for LLVM, but that's a different point).

I'll ponder on how to do this correctly

Reply:
Fixed in https://github.com/rust-lang/rust/pull/67327/files#r357991454